Src/LiquidProjections.NHibernate/NHibernateEventMapConfigurator.cs

```diff
@@ -0,0 +1,246 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace LiquidProjections.NHibernate
+{
+    internal sealed class NHibernateEventMapConfigurator<TProjection, TKey>
+        where TProjection : class, IHaveIdentity<TKey>, new()
+    {
+        private readonly IEventMap<NHibernateProjectionContext> map;
+        private readonly IEnumerable<INHibernateChildProjector> children;
+
+        public NHibernateEventMapConfigurator(
+            IEventMapBuilder<TProjection, TKey, NHibernateProjectionContext> mapBuilder,
+            IEnumerable<INHibernateChildProjector> children = null)
+        {
+            if (mapBuilder == null)
+            {
+                throw new ArgumentNullException(nameof(mapBuilder));
+            }
+
+            map = BuildMap(mapBuilder);
+            this.children = children?.ToList() ?? new List<INHibernateChildProjector>();
+        }
+
+        private IEventMap<NHibernateProjectionContext> BuildMap(
+            IEventMapBuilder<TProjection, TKey, NHibernateProjectionContext> mapBuilder)
+        {
+            mapBuilder.HandleCustomActionsAs((_, projector) => projector());
+
+            mapBuilder.HandleProjectionModificationsAs(async (key, context, projector, options) =>
+            {
+                switch (options.MissingProjectionBehavior)

```

Comment:
Or we could avoid converting back to different methods. We could have a universal method instead. But that method would be less optimal as it would have to always try to load the projection first.